### PR TITLE
[Results] Use unsafeBitcast instead force-cast when accessing elements

### LINKED ISSUE #2844: Swift dynamic objects access throws exception
Hey Realm folks!

I'm running into an exception when trying to access dynamic objects in a Result object:

```swift
class TestObject: Object {
    dynamic var id = 1
}

let realm = try! Realm()
realm.beginWrite()
let test = TestObject()
realm.add(test)
try! realm.commitWrite()
realm.dynamicObjects("TestObject").first  /// Throws an exception
```

The exception being 

```
Could not cast value of type 'Tests.TestObject' (0x7fad72535cc0) to 'RealmSwift.DynamicObject' (0x10bb57060).
```

It looks to be from these lines in `Result`:

```swift
    /**
    Returns the object at the given `index`.

    - parameter index: The index.

    - returns: The object at the given `index`.
    */
    public subscript(index: Int) -> T {
        get {
            throwForNegativeIndex(index)
            return rlmResults[UInt(index)] as! T
        }
    }

    /// Returns the first object in the results, or `nil` if empty.
    public var first: T? { return rlmResults.firstObject() as! T? }

    /// Returns the last object in the results, or `nil` if empty.
    public var last: T? { return rlmResults.lastObject() as! T? }
```

The `as! T?` here is causing the problem, because anything that subclasses Objects (all Realm objects basically) can't be cast with `as!` to DynamicObject, because DynamicObject (also a subclass) is unrelated to those subclasses.  the only as! that would probably work is to Object.

I looked into tests for dynamicObjects and could only find the one:

```swift
    func testDynamicObjects() {
        try! Realm().write {
            try! Realm().create(SwiftIntObject.self, value: [100])
            try! Realm().create(SwiftIntObject.self, value: [200])
            try! Realm().create(SwiftIntObject.self, value: [300])
        }

        XCTAssertEqual(0, try! Realm().dynamicObjects("SwiftStringObject").count)
        XCTAssertEqual(3, try! Realm().dynamicObjects("SwiftIntObject").count)
        assertThrows(try! Realm().dynamicObjects("Object"))
    }
```
Count here actually works.  So I couldn't find any test for actually accessing dynamic objects from a Results<DynamicObject>.

Is this out of scope for DynamicObjects?  Or is this a bug?

### PR DESCRIPTION
**Pull request by mrackwitz**
Fixes #2844. 
/c @bigfish24

### CODE REVIEW

**Comment by tgoyne** [comment]
What type is it bitcasting from in the dynamic case? RLMDynamicObject?

**Comment by mrackwitz** [comment]
That depends on the schema: if there is an accessor class other than `RLMDynamicObject`, which was the case for the filed bug report, then this will be used instead.

**Comment by jpsim** [comment]
I'd like to see more tests for different types, especially `RealmOptionalType`s.

**Comment by mrackwitz** [comment]
Why `RealmOptionalType` specifically? `RealmOptional` isn't really involved in anyway when accessing values through subscripts.

**Comment by jpsim** [comment]
Just testing a single required `Int` property is not sufficient to verify that this works as intended. We should be testing all our supported types, including optionals.

**Comment by mrackwitz** [comment]
@jpsim: Addressed that.

File: RealmSwift-swift1.2/Tests/TestCase.swift [hunk 1]
```diff
@@ -23,6 +23,18 @@ import Realm.Dynamic
...
 import XCTest
 
+func XCTAssertEqual<T: Any where T: Equatable>(@autoclosure a: () -> T?, @autoclosure b: () -> T?, _ message: String? = nil, file: String = __FILE__, line: UInt = __LINE__) {
+    if let _a = a() {
+        if let _b = b() {
+            XCTAssertEqual(_a, _b, (message != nil ? message! : ""), file: file, line: line)
+        } else {
+            XCTFail((message != nil ? message! : "a != nil, b == nil"), file: file, line: line)
+        }
+    } else if let _b = b() {
```
Comment (jpsim) [inline RealmSwift-swift1.2/Tests/TestCase.swift]: don't you get a compiler warning here for not using `_b`? Shouldn't this be `_ = b()` instead?

**Comment by jpsim** [comment]
:+1:

File: RealmSwift-swift1.2/Tests/TestCase.swift (hunk 1 above)
Comment (mrackwitz) [inline RealmSwift-swift1.2/Tests/TestCase.swift]: No, no compiler warning, but good point.

**Pull request merged**